File: 141-LinkedListCycle.js
/*
My solution
O(n) time
O(n) extra memory tehcnically
Iterate through each node and mark as visited. If we find a ndoe we already visited, return true.
If we find a null value at the next node, return false. 
*/
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function (head) {
  let curNode = head;
  while (curNode) {
    //if there is no next node, clearly no loop so we are done
    if (curNode.next === undefined) {
      return false;
    }

    //A next node existed. If we've never visitied current node, mark visited and move on
    if (curNode.checked === undefined) {
      curNode.checked = true;
      curNode = curNode.next;
    } else {
      return true;
    }
  }
  return false;
};

/* Leetcode solution
O(n) time
O(1) memory
Floyd's Cycle Finding Algorithm

*/

var hasCycleLeetcode = function (head) {
    let fast = head;
     while (fast && fast.next) {
       head = head.next;
       fast = fast.next.next;
       if (head === fast) return true;
     }
     return false;
   };
/*
Solution 2: use hash map to mark visited nodes. Can use memory adress 
O(n) time 
O(n) memory
class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        nodes_seen = set()
        while head is not None:
            if head in nodes_seen:
                return True
            nodes_seen.add(head)
            head = head.next
        return False
*/

File: 141-LinkedListCycle.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


''''
My solution
Using hashtable to store memory addresses of each node
O(n) time
O(n) memory
'''
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        dic = {}
        currNode = head
        while (currNode):
            if id(currNode) in dic:
                return True
            else:
                dic[id(currNode)] = 1
                currNode = currNode.next
        return False

''''
Floyd's Cycle Finding Algorithm (Leetcode Solution)
The idea here is to have 2 pointers, 1 pointer which moves at double the pace of the other. Eventually,
if there is indeed a cycle, the two will intersect. If there is no cycle, then we reach a null node which 
will break the loop

O(n) time
O(1) memory
'''
class Solution2:
    def hasCycle(self, head: ListNode) -> bool:
        if head is None:
            return False
        slow = head
        fast = head.next
        while slow != fast:
            if fast is None or fast.next is None:
                return False
            slow = slow.next
            fast = fast.next.next
        return True
